espei/parameter_selection/fitting_steps.py

```diff
@@ -0,0 +1,323 @@
+                    values[..., config_idx] += getattr(fixed_model.endmember_reference_model, cls.data_types_read)
+                else:
+                    pass
+            total_response.append(values.flatten())
+        return total_response
+
+    @classmethod
+    def get_response_vector(cls, fixed_model: Model, fixed_portions: [symengine.Basic], data: [Dataset], sample_condition_dicts: [Dict[str, Any]]) -> ArrayLike:  # np.float_
+        mole_atoms_per_mole_formula_unit = fixed_model._site_ratio_normalization
+        # This function takes Dataset objects (`data`) -> values array (of np.object_)
+        rhs = np.concatenate(cls.shift_reference_state(data, fixed_model), axis=-1)
+        rhs = cls.transform_data(rhs, fixed_model)
+
+        # RKM models are already liner in the parameters, so we don't need to
+        # do anything too special here.
+        # subtract off lower order contributions.
+        for i in range(rhs.shape[0]):
+            rhs[i] -= getattr(fixed_model, cls.parameter_name)
+            if cls.normalize_parameter_per_mole_formula:
+                # Convert the quantity per-mole-atoms to per-mole-formula
+                rhs[i] *= mole_atoms_per_mole_formula_unit
+
+        # Previous steps may have introduced some symbolic terms.
+        # Now we remove all the symbols:
+        # Construct flattened list of site fractions corresponding to the ravelled data
+        site_fractions = []
+        for ds in data:
+            for _ in ds['conditions']['T']:
+                sf = build_sitefractions(fixed_model.phase_name, ds['solver']['sublattice_configurations'], ds['solver'].get('sublattice_occupancies', np.ones((len(ds['solver']['sublattice_configurations']), len(ds['solver']['sublattice_configurations'][0])), dtype=np.float_)))
+                site_fractions.append(sf)
+        site_fractions = list(itertools.chain(*site_fractions))
+        # If any site fractions show up in our rhs that aren't in these
+        # datasets' site fractions, set them to zero. This can happen if we're
+        # fitting a multi-component model that has site fractions from
+        # components that aren't in a particular dataset
+        for sf, value_with_symbols, cond_dict in zip(site_fractions, rhs, sample_condition_dicts):
+            missing_variables = symengine.S(value_with_symbols).atoms(v.SiteFraction) - set(sf.keys())
+            sf.update({x: 0. for x in missing_variables})
+            sf.update(cond_dict)
+        # also replace with database symbols in case we did higher order fitting
+        rhs = [fixed_model.symbol_replace(symengine.S(value_with_symbols).xreplace(sf), fixed_model._symbols).evalf() for value_with_symbols, sf in zip(rhs, site_fractions)]
+        # cast to float, confirming that these are concrete values with no sybolics
+        rhs = np.asarray(rhs, dtype=np.float_)
+        return rhs
+
+
+class StepHM(FittingStep):
```

Comment:
Are there other properties like HM, SM or CPM where the data is linear to the feature and the feature is just a transformation of expr (and expr corresponds to the model attribute)? If there is, do you think it'll make sense to have something similar to AbstractLinearFittingStep (maybe AbstractNonLinearFittingStep) where the user just has to implement transform_feature and possibly transform_data?

I don't know if doing this could obfuscate things and cause the user to think that anything can be fitted with either AbstractLinearFittingStep or AbstractNonLinearFittingStep. When I was fitting directly to the mobility models (rather than fitting to the RK polynomial portion of the mobility), I realized I had to do two different transformation (expr -> linear expression to dataset and model attribute -> linear expression to dataset since the model attribute and expr was different) which I probably only figured out after understanding what StepHM was doing. So it might just be better to leave it so if the user is fitting to something more complicated, they would be able to look at StepHM as a reference to figure out what to do for non-linear properties.